RBD: check for volume group existence

iPraveenParihar opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

This PR adds a check for Volume Group existence in `generateVolumeGroup` call.

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next major release.
* [x] Documentation has been updated, if necessary.
* [x] Unit tests have been added, if necessary.
* [x] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

nixpanic commented on internal/rbd/group/util.go:
```diff
@@ -85,7 +85,10 @@ func (cvg *commonVolumeGroup) generateVolumeGroup(csiID util.CSIIdentifier) erro
 
-	return nil
+	// Check if the volume group exists, by fetching the volume group attributes.
+	_, err = cvg.getVolumeGroupAttributes(context.Background())
```
move this to a new `Exists(context.Context) (bool, error)` function in the VolumeGroup interface, and call it only when/where needed

iPraveenParihar replied:
Thanks!, for the early review 😀

iPraveenParihar replied:
Done

nixpanic commented on internal/util/errors.go:
```diff
@@ -66,5 +72,5 @@ func ShouldRetryVolumeGeneration(err error) bool {
 		errors.Is(err, ErrPoolNotFound) ||
 		errors.Is(err, ErrImageNotFound) ||
-		errors.Is(err, rados.ErrPermissionDenied))
+		errors.Is(err, rados.ErrPermissionDenied) || errors.Is(err, ErrRBDGroupNotFound))
```
please keep the coding style the same, add the new check on its own line

Rakshith-R commented on internal/util/errors.go:
```diff
@@ -66,5 +72,6 @@ func ShouldRetryVolumeGeneration(err error) bool {
 		errors.Is(err, ErrImageNotFound) ||
-		errors.Is(err, rados.ErrPermissionDenied))
+		errors.Is(err, rados.ErrPermissionDenied) ||
+		errors.Is(err, ErrRBDGroupNotFound))
```
Why is group not found error here?

Please create another function for retrying volgroup generation

nixpanic commented on internal/util/errors.go:
```diff
@@ -40,6 +42,10 @@ var (
 	ErrClusterIDNotSet = errors.New("clusterID must be set")
 	// ErrMissingConfigForMonitor is returned when clusterID is not found for the mon.
 	ErrMissingConfigForMonitor = errors.New("missing configuration of cluster ID for monitor")
+	// ErrRBDGroupNotFound is returned when group name is not found in the cluster on the given pool and/or namespace.
```
The `internal/util` package should ideally not have references to RBD, CephFS or NFS. There needs to be a good explanation and advantage to do so. I do not see neither at the moment.

Possibly a `internal/rbd/errors` package would be more appropriate, you could collect all RBD errors there? But it really depends on the reason why you introduced this change.

nixpanic commented on internal/rbd/group/util.go:
```diff
@@ -85,7 +85,9 @@ func (cvg *commonVolumeGroup) generateVolumeGroup(csiID util.CSIIdentifier) erro
 	cvg.pool = pool
 
-	return nil
+	_, err = cvg.Exists(context.Background())
```
Don't call this here, call it at a location that wants to check if the volume group exists.

Rakshith-R replied:
> Don't call this here, call it at a location that wants to check if the volume group exists.

@iPraveenParihar @nixpanic 
I don't think we need to have a function Exists at all. Its called at only two places, that too only after generateVolGroup call 

https://github.com/ceph/ceph-csi/blob/9eb4f61d015696171986f6167a512658d09aec2c/internal/rbd/group/util.go#L474-L482

nixpanic commented on internal/rbd/group/util.go:
```diff
@@ -417,3 +419,13 @@ func (cvg *commonVolumeGroup) GetCreationTime(ctx context.Context) (*time.Time,
 }
+
+// Exists checks if the volume group exists in the journal.
+func (cvg *commonVolumeGroup) Exists(ctx context.Context) (bool, error) {
```
Add this function to the VolumeGroup interface, so that any caller can use it.

nixpanic commented on internal/util/errors.go:
```diff
@@ -66,5 +72,31 @@ func ShouldRetryVolumeGeneration(err error) bool {
+// This helper function is used in scenarios where multiple attempts may be made
+// to retrieve or generate volumegroup information, and we want to gracefully handle
+// specific failure cases while retrying for others.
+func ShouldRetryVolumeGroupGeneration(err error) bool {
```
add a unit-test for this please

nixpanic commented on internal/util/errors.go:
```diff
@@ -66,5 +72,31 @@ func ShouldRetryVolumeGeneration(err error) bool {
+// It checks if the given error matches any of the following known errors:
+//   - ErrPoolNotFound: The rbd pool where the volumegroup/omap is expected doesn't exist.
+//   - ErrRBDGroupNotFound: The volumegroup doesn't exist in the rbd pool.
+//   - rados.ErrPermissionDenied: Permissions to access the pool is denied.
```
Why retry on PermissionDenied? This usually is a pretty final error.

iPraveenParihar replied:
The same check is being done for RBD volumes - _fyi: https://github.com/ceph/ceph-csi/commit/88b964fe18d37af5b53cea00e8d785159fb61f59_

nixpanic replied:
Thanks for the pointer. That commit message contains a description of the _why_, ideally that is included in the comment, but in the commit message it is ok too.

iPraveenParihar committed:
cleanup: move internal/rbd/errors.go to internal/rbd/errors pacakge

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
rbd: check for volume group existence

Signed-off-by: Praveen M <m.praveen@ibm.com>

nixpanic commented:
@iPraveenParihar , the new `Exists()` function in the interface causes the **golangci-lint** to fail. As all interfaces have a `Destroy()` function, you could move that one to a base interface (`journalledObject` maybe) in a separate commit.

iPraveenParihar committed:
cleanup: move Destroy() method to journalledObject interface

VolumeGroup interface has more than 10 methods and it causes
golangci-lint to fail. Moving the `Destroy()` method to a base
interface journalledObject.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
Waiting on @Nikhil-Ladha, for the testing result.

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.30

Nikhil-Ladha commented:
> Waiting on @Nikhil-Ladha, for the testing result.

Tested with different pool ids on both the clusters (and the cluster had other pools too) and, everything looks good.
We can probably optimize the code to avoid dual checks for omap data, but it can be a part of different PR as well.

iPraveenParihar commented:
> Tested with different pool ids on both the clusters (and the cluster had other pools too) and, everything looks good. We can probably optimize the code to avoid dual checks for omap data, but it can be a part of different PR as well.

Thanks @Nikhil-Ladha for the testing! I could think of using the `rbd_group_get_id` from [ceph/ceph](https://github.com/ceph/ceph/blob/812add83d5c135d230f5fc12b41dfa20a5d1dfa1/src/librbd/librbd.cc#L7124C16-L7124C32) but not yet available in go-ceph for now. We could pick it up once we have it in go-ceph.

Madhu-1 commented on internal/csi-addons/rbd/encryptionkeyrotation.go:
```diff
@@ -21,6 +21,7 @@ import (
 	"errors"
 
 	"github.com/ceph/ceph-csi/internal/rbd"
+	rbd_errors "github.com/ceph/ceph-csi/internal/rbd/errors"
```
lets not use `_` use `rbderrors`

Madhu-1 commented on internal/rbd/errors/errors.go:
```diff
@@ -55,4 +63,8 @@ var (
 	ErrImageInUse = errors.New("image is in use")
+	// ErrRBDGroupNotConnected is returned when the RBD group is not connected.
+	ErrRBDGroupNotConnected = fmt.Errorf("%w: RBD group is not connected", librados.ErrNotConnected)
+	// ErrRBDGroupNotFound is returned when group name is not found in the cluster on the given pool and/or namespace.
```
```suggestion
// ErrRBDGroupNotFound is returned when group is not found in the cluster on the given pool and/or namespace.
```

Madhu-1 reviewed (dismissed):
LGTM

Rakshith-R commented on internal/rbd/group/util.go:
```diff
@@ -125,18 +133,27 @@ func (cvg *commonVolumeGroup) generateVolumeGroupFromMapping(
-					if util.ShouldRetryVolumeGeneration(err) {
+					volumeGroupGenerationError = cvg.generateVolumeGroup(mcsiID)
+					if volumeGroupGenerationError == nil {
+						// Check if the volume group exists in the journal.
+						volumeGroupExists, volumeGroupExistsError = cvg.Exists(ctx)
+						if !errors.Is(volumeGroupExistsError, rbderrors.ErrRBDGroupNotFound) {
+							return volumeGroupExistsError
+						}
+					}
+					// If Volume Group doesn't exists or the error is a retryable error,
+					// we should try to get the cluster mapping and generate the volume group from the mapping.
+					if !volumeGroupExists || ShouldRetryVolumeGroupGeneration(volumeGroupGenerationError) {
 						continue
 					}
 
-					return err
+					return volumeGroupGenerationError
 				}
 			}
 		}
 	}
 
-	return util.ErrPoolNotFound
+	return volumeGroupGenerationError
```
I think this is too complex.
Can you try to simplify it ?

Rakshith-R replied:
@iPraveenParihar Can't Exists() be called inside generatevolgroup ? 

I think the Exists logic can be fit inside generateVolGroup

iPraveenParihar replied:
I think @nixpanic had asked to move it out from genereateVolumeGroup  - https://github.com/ceph/ceph-csi/pull/5184#discussion_r1975320212

iPraveenParihar replied:
@Rakshith-R, I tried making it simpler by moving the group exist check code to `generateVolumeGroup` function.

nixpanic commented on internal/rbd/errors/errors.go:
```diff
@@ -55,4 +63,8 @@ var (
 	// ErrImageInUse is returned when the image is in use.
 	ErrImageInUse = errors.New("image is in use")
+	// ErrRBDGroupNotConnected is returned when the RBD group is not connected.
+	ErrRBDGroupNotConnected = fmt.Errorf("%w: RBD group is not connected", librados.ErrNotConnected)
```
This is in a RBD sub-package, drop the RBD part of all variables, make it `ErrGroupNotConnected`.

nixpanic commented on internal/rbd/types/group.go:
```diff
@@ -37,15 +37,15 @@ type journalledObject interface {
 	// GetClusterID returns the ID of the cluster of the object.
 	GetClusterID(ctx context.Context) (string, error)
+
+	// Destroy frees the resources used by the VolumeGroup.
```
VolumeGroup => object

Rakshith-R reviewed (dismissed):
LGTM, Thanks

nixpanic commented on internal/csi-addons/rbd/encryptionkeyrotation.go:
```diff
@@ -21,6 +21,7 @@ import (
 	"errors"
 
 	"github.com/ceph/ceph-csi/internal/rbd"
+	rbderrors "github.com/ceph/ceph-csi/internal/rbd/errors"
```
maybe it makes sense to move common RBD errors to `internal/rbd/types/errors.go`? That is what I decided to do with #5221.

iPraveenParihar replied:
IMO, `rbd/errors` is more suitable than `rbd/types` _(which should be just for the interface and struct definition?)_

nixpanic replied:
That is fine with me too. I just wanted to locate the errors that are documented at the types/interfaces together. I' can move it to `internal/rbd/errors` too.

nixpanic commented on internal/rbd/group/util.go:
```diff
@@ -86,16 +86,25 @@ func (cvg *commonVolumeGroup) generateVolumeGroup(csiID util.CSIIdentifier) erro
 	cvg.pool = pool
 
+	// Check if the volume group exists in the journal.
+	_, err = cvg.getVolumeGroupAttributes(ctx)
```
Does this need to be called here? The initialization of the struct does not need to guarantee that the volume group is recorded in the journal.

iPraveenParihar replied:
I tried to keep it similar to what we have for _[generateVolumeFromVolumeID](https://github.com/ceph/ceph-csi/blob/91baf7681d84fdd450648004764ee1bea419207d/internal/rbd/rbd_util.go#L1145)_.  Here to, we initialise the rbdVol struct and fetch details from journal. Calling outside, made it little complex for readability.

nixpanic replied:
Ok. I just wonder about possible side-effects in case a volumegroup does not exist, or was in the process of getting deleted. I think it should be fine, but maybe you can check that?

iPraveenParihar replied:
>  just wonder about possible side-effects in case a volumegroup does not exist

`initCommonVolumeGroup` is only been called after a VolumeGroup is created, we are good here.

> was in the process of getting deleted

operation locks will prevent this case, but actually I don't see operationa locks on `/DeleteVolumeGroup` & `/CreateVolumeGroup`, was that missed or not required at all?

Nikhil-Ladha replied:
> initCommonVolumeGroup is only been called after a VolumeGroup is created, we are good here.

Actually, we can. For example, if we try to delete an already deleted (or non-existent) volume. Though, looking at the code it would still fail with the same error i.e, `ErrGroupNotFound` but after going through some more function calls.

iPraveenParihar replied:
Yes, that should be fine.

nixpanic commented on internal/rbd/group/util.go:
```diff
@@ -86,16 +86,25 @@ func (cvg *commonVolumeGroup) generateVolumeGroup(csiID util.CSIIdentifier) erro
 
 // generateVolumeGroupFromMapping checks the clusterID and poolID mapping and
 // generates commonVolumeGroup structure for the mapped clusterID and poolID.
+// If the mapping is not found, it returns ErrRBDGroupNotFound.
```
the comment still contains _ErrRBDGroupNotFound_ -> _ErrGroupNotFound_

nixpanic commented:
@mergifyio rebase

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.32

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.30

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.32

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.31

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.30

nixpanic commented:
/test ci/centos/k8s-e2e-external-storage/1.32

nixpanic commented:
/test ci/centos/k8s-e2e-external-storage/1.31

nixpanic commented:
/test ci/centos/k8s-e2e-external-storage/1.30

nixpanic commented:
/test ci/centos/upgrade-tests
